### PR TITLE
Change cloud.Cloud.Regions to slice

### PR DESCRIPTION
**Pull request by axw**

### CODE REVIEW

File: cloud/clouds.go [hunk 1]
```diff
@@ -45,41 +47,78 @@ const (
...
 
 	// StorageEndpoint is the default storage endpoint for the cloud
 	// regions, may be overridden by a region.
-	StorageEndpoint string `yaml:"storage-endpoint,omitempty"`
+	StorageEndpoint string
 
 	// Regions are the regions available in the cloud.
-	Regions map[string]Region `yaml:"regions,omitempty"`
+	//
+	// Regions is a slice, and not a mpa, because order is important:
```
Comment (wallyworld) [inline cloud/clouds.go]: mpa typo

File: cloud/clouds.go [hunk 2]
```diff
@@ -134,28 +173,91 @@ func PublicCloudMetadata(searchPath ...string) (result map[string]Cloud, fallbac
 	return clouds, true, err
 }
 
+var logger = loggo.GetLogger("juju.cloud")
```
Comment (wallyworld) [inline cloud/clouds.go]: move to top?

File: cloud/clouds.go [hunk 3]
```diff
@@ -134,28 +173,91 @@ func PublicCloudMetadata(searchPath ...string) (result map[string]Cloud, fallbac
...
+	// Translate to the exported type. For each cloud, we store
+	// the first region for the cloud as its default region.
+	clouds := make(map[string]Cloud)
+	for name, cloud := range metadata.Clouds {
+		var regions []Region
+		if len(cloud.Regions.Map) > 0 {
+			logger.Debugf("%v", cloud.Regions)
+			for _, item := range cloud.Regions.Slice {
+				name := fmt.Sprint(item.Key)
+				r := cloud.Regions.Map[name]
```
Comment (wallyworld) [inline cloud/clouds.go]: why can't we use item.Value here?
I don't get why we need both a Map and a Slice

File: cloud/clouds.go [hunk 4]
```diff
@@ -45,41 +47,78 @@ const (
...
+//
+// When marshalling, we populate the Slice field only. This is
+// necessary for us to control the order of map items.
+//
+// When unmarshalling, we populate both Map and Slice. Map is
+// populated to simplify conversion to Region objects. Slice
+// is populated so we can identify the first map item, which
+// becomes the default region for the cloud.
+type regions struct {
+	Map   map[string]*region
```
Comment (wallyworld) [inline cloud/clouds.go]: there's no yaml tags here - are Map and Slice implicitly support by goyaml.v2?

File: cloud/clouds.go (hunk 2 above)
Comment (axw) [inline cloud/clouds.go]: will be deleting, just here while debugging an issue

File: cloud/clouds.go (hunk 4 above)
Comment (axw) [inline cloud/clouds.go]: using MarshalYAML, don't need tags

File: cloud/clouds.go (hunk 3 above)
Comment (axw) [inline cloud/clouds.go]: This means I don't have to extract map items explicitly, so easier to maintain (Region may grow fields). The MapSlice is used only for the key ordering.

Comment (wallyworld) [inline cloud/clouds.go]: r := item.Value.(*region)

won't work?

		var regions []Region
 		if len(cloud.Regions.Slice) > 0 {
 			for _, item := range cloud.Regions.Slice {
 				name := fmt.Sprint(item.Key)
 				r := item.Value.(*region)
				regions = append(regions, Region{
 					name, r.Endpoint, r.StorageEndpoint,
 				})
 			}
 		}

File: cloud/clouds.go (hunk 1 above)
Comment (axw) [inline cloud/clouds.go]: fixed

**Comment by axw** [comment]
Retargeting to master.

**Pull request closed without merging**